src/TracingMiddleware/TracingMiddleware.cs

```diff
@@ -3,88 +3,107 @@
+        {
+            this.nextFunc = nextFunc;
+            this.tracer = tracer;
         }
 
-        public static MidFunc Tracing(ITracer tracer)
+        public async Task Invoke(HttpContext context)
         {
-            if (tracer == null)
-                throw new ArgumentNullException("tracer");
-            tracer = new SafeTracer(tracer);
-            return next => async env =>
+            this.tracer = new SafeTracer(this.tracer);
+            if (!tracer.IsEnabled)
             {
-                if (!tracer.IsEnabled)
+                await this.nextFunc(context).ConfigureAwait(false);
+            }
+            else
+            {
+                string requestId;
+                if (!string.IsNullOrWhiteSpace(context.TraceIdentifier))
                 {
-                    await next(env).ConfigureAwait(false);
+                    requestId = context.TraceIdentifier;
                 }
                 else
                 {
-                    string requestId;
-                    if (env.ContainsKey("owin.RequestId"))
+                    requestId = Guid.NewGuid().ToString();
+                    context.TraceIdentifier = requestId;
+                }
+
+                var path = context.Request.Path + context.Request.QueryString;
+                tracer.Trace(requestId, "Request Start: " + path);
+
+                var stopWatch = Stopwatch.StartNew();
+
+                try
+                {
+                    await this.nextFunc(context).ConfigureAwait(false);
+                }
+                catch (Exception exception)
+                {
+                    tracer.Trace(requestId, exception.ToString());
+                    throw;
+                }
+                finally
+                {
+                    stopWatch.Stop();
+
+                    if (tracer.Filters.All(filter => filter.Invoke(context)))
                     {
-                        Guid result;
-                        if (Guid.TryParse(env["owin.RequestId"].ToString(), out result) && result != Guid.Empty)
+                        this.tracer.Trace(requestId, $"Request {nameof(context.Request.Cookies)}", context.Request.Cookies);
+                        this.tracer.Trace(requestId, $"Request {nameof(context.Request.Host)}", context.Request.Host);
+                        this.tracer.Trace(requestId, $"Request {nameof(context.Request.ContentType)}", context.Request.ContentType);
+                        this.tracer.Trace(requestId, $"Request {nameof(context.User)}", context.User);
+                        this.tracer.Trace(requestId, $"Request {nameof(context.Request.Cookies)}", context.Request.Cookies);
+
+                        var req = context.Features.Get<IHttpRequestFeature>();
+                        if (req != null)
                         {
-                           requestId = result.ToString();   
+                            this.tracer.Trace(requestId, $"Request {nameof(req.Method)}", req.Method);

```

Comment:
Wait, some of the request information is in the context, some you have to pull it out of this feature thing? What?